MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java

```diff
@@ -977,48 +982,96 @@ public static Entity getInfantryEntity(UnitGeneratorParameters params, int skill
 
                 UnitGeneratorParameters newParams = params.clone();
                 newParams.clearMovementModes();
-                newParams.setUnitType(UnitType.INFANTRY);
                 newParams.setWeightClass(AtBDynamicScenarioFactory.UNIT_WEIGHT_UNSPECIFIED);
-
-                // to save ourselves having to re-generate a bunch of infantry for smaller bays (3 tons and lower)
-                // we will limit ourselves to generating low-weight foot platoons
-                if(bayCapacity <= IUnitGenerator.FOOT_PLATOON_INFANTRY_WEIGHT) {
-                    newParams.getMovementModes().add(EntityMovementMode.INF_LEG);
-                    newParams.setFilter(inf -> inf.getTons() <= IUnitGenerator.FOOT_PLATOON_INFANTRY_WEIGHT);
-                } else {
-                    newParams.getMovementModes().addAll(IUnitGenerator.ALL_INFANTRY_MODES);
-                    newParams.setFilter(inf -> inf.getTons() <= bayCapacity);
+                
+                Entity transportedUnit = null;
+                
+                // for now, we'll assign BA units with greater likelyhood to units with higher-rated equipment
+                int baRoll = Compute.d6(2);                
+                if(baRoll >= infantryToBAUpgradeTNs[params.getQuality()]) {
+                    transportedUnit = generateTransportedBAUnit(newParams, bayCapacity, skill, campaign);
+                }
+                
+                if(transportedUnit == null) {
+                    transportedUnit = generateTransportedInfantryUnit(newParams, bayCapacity, skill, campaign);
                 }
 
-                MechSummary ms = campaign.getUnitGenerator().generate(newParams);
+                // sometimes something crazy will happen and we will not be able to load the unit into the transport
+                // so let's at least have it deploy at the same time as the transport
+                transportedUnit.setDeployRound(transport.getDeployRound());
```

Comment:
Is `transportedUnit` guaranteed to be non-null here?

Reply:
Probably not. I'll bullet-proof it.